### PR TITLE
Don't show edit-or-discard-answers modal if applicant hasn't started answering a block

### LINKED ISSUE #6987: Don't show "edit or discard answers" modal if no previously filled in answers and no filled in answers now
**Describe the bug**
As part of https://github.com/civiform/civiform/issues/6450, we updated the "Review" and "Previous" buttons to save user answers. If the answers fail validation (e.g. the applicant didn't answer a required question), then we display a modal asking the user to edit their answers or proceed without saving -- see https://github.com/civiform/civiform/blob/main/browser-test/image_snapshots/application_navigation_test/error-on-review-modal.png.

This modal will *also* pop up if you open a block that you haven't seen before, don't answer any questions, then click "Review" or "Previous". This is pretty unintuitive: I can image applicants opening a block then deciding "eh I don't want to fill this out right now" and wanting to click "Review" or "Previous" to see what they've done so far. We shouldn't pop up a modal, we should just let them proceed.

Something we should **avoid** doing: If a user has previously filled in a block, then opens it and deletes their answers to required questions, we **should** pop up the modal.

**To Reproduce**

1. Enable SAVE_ON_ALL_ACTIONS flag.
2. As an applicant, apply to a program and open a block with required questions.
3. Don't answer any.
4. Click "Review" or "Previous".

**Actual behavior**
The "edit or discard answers" modal appears.

**Expected behavior**
No modal appears, and I'm taken directly to the review page or previous page (depending on what button I clicked).

### PR DESCRIPTION
**Pull request by caitlinshk**
### Description

As part of https://github.com/civiform/civiform/issues/6450, we updated the "Review" and "Previous" buttons to save user answers. If the answers fail validation (e.g. the applicant didn't answer a required question), then we display a modal asking the user to edit their answers or proceed without saving -- see https://github.com/civiform/civiform/blob/main/browser-test/image_snapshots/application_navigation_test/error-on-review-modal.png.

This modal will also pop up if you open a block that you haven't seen before, don't answer any questions, then click "Review" or "Previous". This is pretty unintuitive: applicants may open a block, decide they don't want to fill it out, and click "Review" or "Previous" to see what they've done so far. We should let them proceed to the review page or previous page and not show the modal.

This PR updates the "Review" and "Previous" buttons to **not** show the modal if the applicant opens a block they haven't answered before and fills nothing out. Instead, the applicant is taken directly to the review page / previous page.

Note: This is **not** protected behind the `SAVE_ON_ALL_ACTIONS` feature flag because `if` block I added only triggers when the `ApplicantRequestedAction` is `REVIEW_PAGE` or `PREVIOUS_BLOCK`, and those actions are never used if the `SAVE_ON_ALL_ACTIONS` flag is off.

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)

- [x] Added the correct label: < feature | enhancement | bug | under-development | dependencies | infrastructure | ignore-for-release | database >
- [ ] Assigned to a specific person, `civiform/developers`, or a [more specific round-robin list](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#adding-reviewers)
- [ ] Added an additional reviewer from outside your organization as FYI (if the primary reviewer is in the same organization as you)
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)

### Instructions for manual testing

Part 1
- Open a block with required questions, don’t answer any
     - click Review [or Previous] -> verify you go to review page [or previous page] without seeing the modal
     - click Save&Next-> verify you stay on the same page and see errors

Part 2
- Open a block with all optional questions, don’t answer any, click Review [or Previous or Save&next], then fill out other blocks but never reopen that block -> verify you can submit the form

Part 3
- Open a block with required questions, fill them out and save. Re-open that block and delete all answers
     - click Review [or Previous] -> verify error modal appears on top of the block and errors are shown underneath the modal
     - click Save&next ->  verify the errors are immediately shown

Part 4
- Open a block with some optional and some required questions, only answer the required questions,  click Review [or Previous] -> verify you go to review page [or previous page] and your answers were saved. Verify you can submit the form.

### Issue(s) this completes

Fixes #6987

### CODE REVIEW

**Comment by caitlinshk** [comment]
* **#6992** <a href="https://app.graphite.dev/github/pr/civiform/civiform/6992?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @caitlinshk and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: server/app/controllers/applicant/ApplicantProgramBlocksController.java [hunk 1]
```diff
@@ -846,11 +904,31 @@ private CompletionStage<Result> renderErrorOrRedirectToRequestedPage(
...
+        blockId,
+        inReview,
+        applicantRequestedAction,
+        roApplicantProgramService,
+        flashingMap);
+  }
 
-    if (applicantRequestedAction == ApplicantRequestedAction.REVIEW_PAGE) {
+  /** Returns the correct page based on the given {@code applicantRequestedAction}. */
+  private CompletionStage<Result> getRequestedPage(
```
Comment (dkatzz) [inline server/app/controllers/applicant/ApplicantProgramBlocksController.java:919]: **nitpick:** Should we name redirectToRequestedPage?

File: server/test/controllers/applicant/ApplicantProgramBlocksControllerTest.java [hunk 2]
```diff
@@ -679,6 +683,374 @@ public void update_withValidationErrors_requestedActionPrevious_staysOnBlockAndS
...
+        .join();
+
+    applicant.refresh();
+    // We mark as question as seen by including the "updated_at" metadata.
+    assertThat(applicant.getApplicantData().asJsonString())
+        .containsPattern("\"applicant_name\":\\{\"updated_at\":.");
+  }
+
+  @Test
+  @Parameters({"NEXT_BLOCK", "REVIEW_PAGE", "PREVIOUS_BLOCK"})
```
Comment (dkatzz) [inline server/test/controllers/applicant/ApplicantProgramBlocksControllerTest.java:887]: **praise:** This setup is nice!

**Review by dkatzz — APPROVED** [review]

File: browser-test/src/applicant/application_navigation.test.ts
[the remark this answers is not in the document]
Comment (gwendolyngoetz) [reply browser-test/src/applicant/application_navigation.test.ts:1]: +1

File: server/app/controllers/applicant/ApplicantProgramBlocksController.java [hunk 3]
```diff
@@ -657,46 +659,102 @@ public CompletionStage<Result> updateWithApplicantId(
...
+                  // [Explanation: We don't allow applicants to submit an application until they've
+                  // at least seen all the questions. We mark as question as "seen" by checking
+                  // that the metadata is filled in (see
+                  // {@link ApplicantQuestion#isAnsweredOrSkippedOptionalInProgram}).
+                  //
+                  // If a block has all optional questions, having all empty answers is a valid
+                  // response and we should mark that block as seen by having
+                  // {@link #stageAndUpdateIfValid} fill in that metadata. So, we can't immediately
+                  // navigate away.]
+                  && !maybeBlockBeforeUpdate.get().hasOnlyOptionalQuestions()) {
```
Comment (gwendolyngoetz) [inline server/app/controllers/applicant/ApplicantProgramBlocksController.java:726]: (blocking) This `if` statement is getting pretty long and complicated. The documentation helps immensely, but please move it into a function of it's own called `canNavigateAwayFromPageImmediately` or something along those lines.

File: server/app/controllers/applicant/ApplicantProgramBlocksController.java (hunk 1 above)
Comment (gwendolyngoetz) [reply server/app/controllers/applicant/ApplicantProgramBlocksController.java]: +1

File: server/test/controllers/applicant/ApplicantProgramBlocksControllerTest.java (hunk 2 above)
Comment (gwendolyngoetz) [reply server/test/controllers/applicant/ApplicantProgramBlocksControllerTest.java:887]: +1

File: server/app/controllers/applicant/ApplicantProgramBlocksController.java [hunk 4]
```diff
@@ -657,46 +659,102 @@ public CompletionStage<Result> updateWithApplicantId(
...
+                  readOnlyApplicantProgramService.getActiveBlock(blockId);
+              ApplicantRequestedAction applicantRequestedAction =
+                  applicantRequestedActionWrapper.getAction();
+
+              // Allow applicants to immediately navigate away from a block if they haven't even
+              // started answering it. We can immediately navigate away from a block if:
+              if (
+              // There are no answers currently filled out...
+              formData.values().stream().allMatch(value -> value.equals(""))
+                  // ... and the applicant wants to navigate to previous or review...
```
[the remark this answers is not in the document]
Comment (caitlinshk) [reply server/app/controllers/applicant/ApplicantProgramBlocksController.java:706]: Great question. Do you know if selecting no options for a required checkbox question is a valid response?

When I just tried it, it seems like that's not It looks like it's not valid? I tried it with the "Which of the following kitchen instruments do you own?" question in the Comprehensive Sample Program by answering everything in the block except that question and then clicking Save&next, but I got the "This question is required" error. I also tried changing that question on the admin side to allow 0 as the minimum number of choices required, but I still got the "This question is required" error.

So, I think that means that this format is correct for all questions? I also just checked that if you have a block that only has static text on it, you can use any of the three buttons and you'll be taken to the right place.

I could add browser tests to each individual question type browser test that test each of the three buttons with and without an answer, but maybe that's overkill :)

File: server/app/controllers/applicant/ApplicantProgramBlocksController.java [hunk 5]
```diff
@@ -657,46 +659,102 @@ public CompletionStage<Result> updateWithApplicantId(
...
+                  // [Explanation: We can't let applicants navigate to the next block without
+                  // answers because the next block may have a visibility conditions that
+                  // depends on the answers to this block.]
+                  && (applicantRequestedAction == REVIEW_PAGE
+                      || applicantRequestedAction == PREVIOUS_BLOCK)
+                  && maybeBlockBeforeUpdate.isPresent()
+                  // ... and the applicant didn't previously complete this block...
```
[the remark this answers is not in the document]
Comment (caitlinshk) [reply server/app/controllers/applicant/ApplicantProgramBlocksController.java:712]: Done.

File: server/app/controllers/applicant/ApplicantProgramBlocksController.java (hunk 1 above)
Comment (caitlinshk) [reply server/app/controllers/applicant/ApplicantProgramBlocksController.java]: Great name! Done.

**Review by gwendolyngoetz — CHANGES_REQUESTED** [review]

File: server/app/controllers/applicant/ApplicantProgramBlocksController.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (dkatzz) [reply server/app/controllers/applicant/ApplicantProgramBlocksController.java]: Got it, thanks for checking! Hmm yeah - it is kind of surprising that we don't allow people to proceed without selecting anything, but maybe not a bug? Probably worth just mentioning under checkbox question [here](https://docs.civiform.us/user-manual/civiform-admin-guide/working-with-questions/manage-questions#question-types) that if you want to allow no options to be selected, the question should be optional and maybe we should make it so we don't allow 0 as the minimum and instead say it should be set as optional if we expect people to not select anything. Either way, out of scope for this change :)

[the remark this answers is not in the document]
Comment (dkatzz) [reply server/app/controllers/applicant/ApplicantProgramBlocksController.java]: And I think the tests are good as is, but if in the future there's a type that we check differently than through that formData check, we can add a specific test for that.

**Review by gwendolyngoetz — APPROVED** [review]

File: server/app/controllers/applicant/ApplicantProgramBlocksController.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (caitlinshk) [reply server/app/controllers/applicant/ApplicantProgramBlocksController.java]: Sounds good! Created https://github.com/civiform/civiform/issues/7067 to track the doc updates to the checkbox question.

**Pull request merged**